collector/ipvs.go

```diff
@@ -0,0 +1,258 @@
+
+type ipvsCollector struct {
+	config                                                           Config
+	backendConnectionsActive, backendConnectionsInAct, backendWeight *prometheus.GaugeVec
+	stats                                                            map[string]prometheus.Counter
+}
+
+type ipvsBackendStatus struct {
+	localAddress  net.IP
+	localPort     uint16
+	proto         string
+	remoteAddress net.IP
+	remotePort    uint16
+	activeConn    uint16
+	inActConn     uint16
+	weight        uint16
+}
+
+func init() {
+	Factories["ipvs"] = NewIpvsCollector
+}
+
+func NewIpvsCollector(config Config) (Collector, error) {
+	c := ipvsCollector{
+		config: config,
+		backendConnectionsActive: &prometheus.GaugeVec{},
+		backendConnectionsInAct:  &prometheus.GaugeVec{},
+		backendWeight:            &prometheus.GaugeVec{},
+		stats:                    map[string]prometheus.Counter{},
+	}
+
+	for _, metricName := range ipvsStatFieldNames {
+		c.stats[metricName] = prometheus.NewCounter(
+			prometheus.CounterOpts{
+				Namespace: Namespace,
+				Subsystem: ipvsSubsystem,
+				Name:      metricName,
+			},
+		)
+	}
+
+	c.backendConnectionsActive = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: Namespace,
+			Subsystem: ipvsSubsystem,
+			Name:      "backend_connections_active",
+		},
+		ipvsBackendLabelNames,
+	)
+	c.backendConnectionsInAct = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: Namespace,
+			Subsystem: ipvsSubsystem,
+			Name:      "backend_connections_inactive",
+		},
+		ipvsBackendLabelNames,
+	)
+	c.backendWeight = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: Namespace,
+			Subsystem: ipvsSubsystem,
+			Name:      "backend_weight",
+		},
+		ipvsBackendLabelNames,
+	)
+
+	return &c, nil
+}
+
+func (c *ipvsCollector) Update(ch chan<- prometheus.Metric) (err error) {
+	ipvsStats, err := getIpvsStats()
+	if err != nil {
+		return fmt.Errorf("could not get IPVS stats: %s", err)
+	}
+
+	for _, metricName := range ipvsStatFieldNames {
+		c.stats[metricName].Set(float64(ipvsStats[metricName]))
+		c.stats[metricName].Collect(ch)
+	}
+
+	backendStats, err := getIpvsBackendStatus()
+	if err != nil {
+		return fmt.Errorf("could not get backend status: %s", err)
+	}
+
+	for _, backend := range backendStats {
+		labelValues := []string{
+			backend.localAddress.String(),
+			strconv.FormatUint(uint64(backend.localPort), 10),
+			backend.remoteAddress.String(),
+			strconv.FormatUint(uint64(backend.remotePort), 10),
+			backend.proto,
+		}
+		c.backendConnectionsActive.WithLabelValues(labelValues...).Set(float64(backend.activeConn))
+		c.backendConnectionsInAct.WithLabelValues(labelValues...).Set(float64(backend.inActConn))
+		c.backendWeight.WithLabelValues(labelValues...).Set(float64(backend.weight))
+	}
+
+	c.backendConnectionsActive.Collect(ch)
+	c.backendConnectionsInAct.Collect(ch)
+	c.backendWeight.Collect(ch)
+
+	return nil
+}
+
+func getIpvsBackendStatus() ([]ipvsBackendStatus, error) {
```

Comment:
It seems like this whole parsing functionality would best live in https://github.com/prometheus/procfs, where all the other procfs-related code is already.